pkg/causality/internal/slots.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"container/list"
+	"sync"
+)
+
+// Eq describes objects that can be compared for equality.
+type Eq[T any] interface {
+	Equals(other T) bool
+}
+
+// Slots implements slot-based conflict detection.
+// It holds references to E, which can be used to build
+// a DAG of dependency.
+type Slots[E Eq[E]] struct {
+	mu    sync.Mutex
+	slots map[int64]*list.List
```

Comment:
Can we use `sync.Map` here?

Reply:
We still need a mutex to make `Add` race-free.

Reply:
Got it! Thanks!